### PR TITLE
RFC: Resize small arrays, Issue #33902

### PR DESCRIPTION
**Pull request by efwright**
Issue #33902 shows a situation in which small arrays can be grown using sizehint! but not shrunken. I am very new to working with Julia, so this is mostly to start to get my bearings. I found that the jl_array_shrink is setup to just ignore arrays that are too small (since small arrays are not malloc'd), but these arrays can be changed by jl_array_grow_end. I tried to use the grow code, and the shrink on the large array code to write something that would allow shrinking small arrays.

### CODE REVIEW

File: test/smallarrayshrink.jl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    @test x[4] == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 0) == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 1) == 4
+    sizehint!(x, 10000)
+    @test x[1] == 1
+    @test x[2] == 2
+    @test x[3] == 3
+    @test x[4] == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 0) == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 1) == 10000
```
Comment (JeffBezanson) [inline test/smallarrayshrink.jl]: This is quite "clever"/obfuscated... took me a while to figure out. This is relying on the fact that `maxsize` is stored after `nrows` and that this function doesn't check its arguments. We would need a different way to do this.

**Review by JeffBezanson — CHANGES_REQUESTED** [review]

File: src/array.c [hunk 2]
```diff
@@ -948,6 +948,23 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
...
         //this is a julia-allocated buffer that needs to be marked
+        char *typetagdata;
+        char *newtypetagdata;
+        if (isbitsunion) {
+            typetagdata = (char*)malloc(a->nrows);
+            memcpy(typetagdata, jl_array_typetagdata(a), a->nrows);
+        }
+        char * originaldata = (char*) a->data - a->offset * a->elsize;
+        char * newdata = (char*) jl_gc_alloc_buf(jl_get_ptls_states(), newbytes);
+        jl_gc_wb_buf(a, a->data, newbytes);
```
Comment (vtjnash) [inline src/array.c]:
```suggestion
```

File: src/array.c [hunk 3]
```diff
@@ -948,6 +948,23 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
...
     if (a->flags.how == 1) {
         //this is a julia-allocated buffer that needs to be marked
+        char *typetagdata;
+        char *newtypetagdata;
+        if (isbitsunion) {
+            typetagdata = (char*)malloc(a->nrows);
+            memcpy(typetagdata, jl_array_typetagdata(a), a->nrows);
+        }
+        char * originaldata = (char*) a->data - a->offset * a->elsize;
+        char * newdata = (char*) jl_gc_alloc_buf(jl_get_ptls_states(), newbytes);
```
Comment (vtjnash) [inline src/array.c]:
```suggestion
        char *newdata = (char*)jl_gc_alloc_buf(jl_get_ptls_states(), newbytes);
```

File: src/array.c [hunk 4]
```diff
@@ -948,6 +948,23 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
     char *originalptr = ((char*) a->data) - a->offset * a->elsize;
     if (a->flags.how == 1) {
         //this is a julia-allocated buffer that needs to be marked
+        char *typetagdata;
+        char *newtypetagdata;
+        if (isbitsunion) {
+            typetagdata = (char*)malloc(a->nrows);
```
Comment (vtjnash) [inline src/array.c]:
```suggestion
            typetagdata = (char*)malloc_s(a->nrows);
```

File: src/array.c [hunk 5]
```diff
@@ -948,6 +948,23 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
...
+            memcpy(typetagdata, jl_array_typetagdata(a), a->nrows);
+        }
+        char * originaldata = (char*) a->data - a->offset * a->elsize;
```
Comment (JeffBezanson) [inline src/array.c]:
```suggestion
        char *originaldata = (char*) a->data - a->offset * a->elsize;
```

File: src/array.c (hunk 3 above)
Comment (JeffBezanson) [inline src/array.c]:
```suggestion
        char *newdata = (char*) jl_gc_alloc_buf(jl_get_ptls_states(), newbytes);
```

File: src/array.c (hunk 4 above)
Comment (JeffBezanson) [inline src/array.c]:
```suggestion
            typetagdata = (char*)malloc_s(a->nrows);
```

File: src/array.c [hunk 6]
```diff
@@ -948,6 +948,23 @@ STATIC_INLINE void jl_array_shrink(jl_array_t *a, size_t dec)
...
+        char * newdata = (char*) jl_gc_alloc_buf(jl_get_ptls_states(), newbytes);
+        jl_gc_wb_buf(a, a->data, newbytes);
+        a->maxsize -= dec;
+        if (isbitsunion) {
+            newtypetagdata = jl_array_typetagdata(a);
+            memcpy(newtypetagdata, typetagdata, a->nrows);
+            free(typetagdata);
+        }
+        memcpy(newdata, originaldata, newbytes);
+        a->data = newdata + a->offset*a->elsize;
```
Comment (vtjnash) [inline src/array.c]:
```suggestion
        a->data = newdata + a->offset * elsz;
```

File: src/array.c (hunk 2 above)
Comment (JeffBezanson) [reply src/array.c]: This needs to be `jl_gc_wb_buf(a, newdata, newbytes)`

**Comment by JeffBezanson** [comment]
Welcome, @efwright ! Thanks for this. Basically right, just needs a couple small things.

File: src/array.c (hunk 2 above)
Comment (vtjnash) [reply src/array.c:959]:
```suggestion
        jl_gc_wb_buf(a, newdata, newbytes);
```

**Pull request merged**